src/app/cmd/set_layer_tileset.cpp

```diff
@@ -39,5 +40,10 @@ void SetLayerTileset::onUndo()
   layer->incrementVersion();
 }
 
+void SetLayerTileset::onFireNotifications()
+{
+  App::instance()->PaletteChange();
+}
```

Comment:
We don't access `App` instance directly from Cmd notifcations:
```suggestion
{
  doc::Sprite* sprite = this->sprite();
  Doc* doc = static_cast<Doc*>(sprite->document());
  doc->notifyPaletteChanged();
}
```
This exact same code is from `SetPalette::onFireNotifications()`.

Reply:
I have tried that when I was working on this and it didn't work. Because I need that PaletteView::onAppPaletteChange gets called, which in turn calls the paletteChange method of the adapter. AFAIK PaletteView::onAppPaletteChange is called only when App::instance()->PaletteChange() is called. Am I missing something?

Btw, just to double check, I tried again with the suggested code and checked that it makes Aseprite crash after changing the tileset of a tilemap and trying to pick a tile in the tile selector.

Reply:
Didn't test the change, probably we need another kind of notification notifyTilesetChanged (which makes more sense because we're changing the tileset, not the palette). Anyway I'm not sure why this was originally required.

Reply:
I see. Then we should take note about updating this function in palettes.cpp, right?:
```c++
// Changes the current system palette and triggers the
// App::PaletteChange signal.
//
// If "_palette" is nullptr the default palette is set.
bool set_current_palette(const Palette *_palette, bool forced)
{
  const Palette* palette = (_palette ? _palette: ase_default_palette);
  bool ret = false;

  // Have changes
  if (forced ||
      palette->countDiff(ase_current_palette, NULL, NULL) > 0) {
    // Copy current palette
    palette->copyColorsTo(ase_current_palette);

    // Call slots in signals
    App::instance()->PaletteChange();  // <--- this call would update the tiles palette too?

    ret = true;
  }

  return ret;
}
```

Reply:
The `set_current_palette` function (and the `App::PaletteChange` signal) is an old function from the Allegro era, were the graphics mode could be in 8-bit (and there were an actual "current palette" set in the screen). Now this function (and almost everything from `modules/` [should be removed](https://github.com/aseprite/aseprite/blob/main/src/app/modules/README.md)/replaced with other kind of notifications).

Right now it's a kind of workaround for some situations but probably `ContextObserver::onActiveSiteChange` will be the correct place to listen if the current palette changes. Anyway this is refactor for the future.

I'm not sure why `set_current_palette` should update the tileset view. What exactly are you trying to achieve and where? If the `PaletteView+TilesetViewAdapter` must be updated because the current tileset was changed, I think `ColorBar::onTilesetChanged` already does that (probably you can generate a `Doc::notifyTilesetChanged()`).

Reply:
Never mind...to my surprise (and maybe yours 😄) there was already a notifyTilesetChanged that you introduced a long time ago, and it seems to work as well. But now I have to fix a bug I found while trying some things.